### PR TITLE
[automatic failover] Implement grace period on failover

### PR DESCRIPTION
**Pull request by atakavci**
# Add Grace Period Support for Multi-Database Failover

## Overview

This PR introduces a **grace period** mechanism for the multi-database failover feature in Lettuce. The grace period prevents immediate failback to a database that was recently failed over from, ensuring system stability and preventing rapid switching between databases (flapping).

## Problem Statement

Without a grace period, when a database becomes unhealthy and triggers a failover, it could immediately become healthy again and trigger a failback. This can lead to:
- Rapid switching between databases (flapping)
- Instability in the system
- Unnecessary connection churn
- Potential cascading failures

## Solution

Implemented a configurable grace period that starts when a database is failed over from. During this period:
- The database cannot be selected for automatic failover or failback
- The circuit breaker state is tracked and reset when the grace period expires
- Forced manual switches (via `switchTo()`) are still allowed, ignoring the grace period
- After the grace period expires, the database becomes eligible for selection again

### 1. **MultiDbOptions Enhancement**
- Added `gracePeriod` configuration option (default: 30 seconds)
- Configurable via `MultiDbOptions.Builder.gracePeriod(long)`
- Can be set to `0` to disable grace period functionality

### 2. **RedisDatabaseImpl Updates**
- Added grace period tracking with `gracePeriodEndTime` field
- New methods:
  - `startGracePeriod(long)` - Starts a grace period for the specified duration
  - `isInGracePeriod()` - Checks if database is currently in grace period
  - `clearGracePeriod()` - Clears the grace period
  - `isHealthy()` - Checks health including grace period status
  - `isHealthyIgnoreGracePeriod()` - Checks health without grace period consideration
- Automatic circuit breaker reset to `CLOSED` when grace period expires

### 3. **StatefulRedisMultiDbConnectionImpl Updates**
- Grace period is started when failover occurs due to:
  - Circuit breaker state change (OPEN/HALF_OPEN)
  - Health check failure
- Grace period is respected during:
  - Automatic failover selection
  - Automatic failback checks
- Grace period is **ignored** for:
  - Forced manual switches via `switchTo()`
- Removed `DatabasePredicates.isHealthyAndCbClosed` in favor of instance methods

### 4. **CircuitBreaker Interface Update**
- Made `transitionTo(State)` method public in the interface
- Allows external control of circuit breaker state (needed for grace period reset)

### Unit Tests
- **RedisDatabaseImplGracePeriodUnitTests**: Comprehensive unit tests for grace period tracking
  - Grace period start/stop behavior
  - Circuit breaker reset on expiry
  - Health check integration
  - Edge cases (zero, negative durations)

- **MultiDbOptionsUnitTests**: Configuration tests
  - Default grace period value
  - Custom grace period configuration
  - Zero grace period (disabled)
  - Builder method chaining

### Integration Tests
- **GracePeriodIntegrationTests**: End-to-end integration tests
  - Manual switch and automatic failback behavior
  - Forced switch ignoring grace period
  - Zero grace period configuration
  - Grace period preventing premature failback
  - Grace period expiry allowing failback

### Test Infrastructure Updates
- Added `@NoFailback` qualifier annotation for test injection
- Updated existing tests to use `@NoFailback` where appropriate to avoid interference from automatic failback
- Added `QualifiedMultiDbClientInjectionExampleTest` demonstrating qualifier usage

## Configuration Example

```java
// Default grace period (30 seconds)
MultiDbOptions options = MultiDbOptions.builder().build();

// Custom grace period (60 seconds)
MultiDbOptions options = MultiDbOptions.builder()
    .gracePeriod(60000L)
    .build();

// Disable grace period
MultiDbOptions options = MultiDbOptions.builder()
    .gracePeriod(0L)
    .build();
```

### Before
- Database becomes unhealthy → failover to backup
- Database becomes healthy again → immediate failback
- Potential for rapid switching

### After
- Database becomes unhealthy → failover to backup + start grace period
- Database becomes healthy during grace period → no failback (stays on backup)
- Grace period expires → database eligible for failback
- Manual `switchTo()` → always allowed, ignores grace period

## Backward Compatibility

- **Fully backward compatible**: Default grace period of 30 seconds is applied
- Existing code continues to work without changes
- Grace period can be disabled by setting it to `0` for legacy behavior

### Core Implementation
- `CircuitBreaker.java` - Added public `transitionTo()` method
- `CircuitBreakerImpl.java` - Made `transitionTo()` public
- `MultiDbOptions.java` - Added grace period configuration
- `RedisDatabaseImpl.java` - Implemented grace period tracking
- `StatefulRedisMultiDbConnectionImpl.java` - Integrated grace period into failover logic

### Tests
- `GracePeriodIntegrationTests.java` - New integration tests
- `RedisDatabaseImplGracePeriodUnitTests.java` - New unit tests
- `MultiDbOptionsUnitTests.java` - Added grace period configuration tests
- `QualifiedMultiDbClientInjectionExampleTest.java` - New example test
- Updated existing tests to use `@NoFailback` annotation where needed

## Related Issues

This enhancement improves the stability and reliability of the multi-database failover feature by preventing database flapping.

### CODE REVIEW

File: src/main/java/io/lettuce/core/failover/MultiDbOptions.java [hunk 1]
```diff
@@ -65,11 +78,16 @@ public static class Builder {
...
 
+        /** Default grace period duration. */
+        private static final long GRACE_PERIOD_DEFAULT = 30000;
+
         /** Whether automatic failback to higher-priority databases is supported. */
         private boolean failbackSupported = true;
 
         private long failbackCheckInterval = FAILBACK_CHECK_INTERVAL_DEFAULT;
 
+        private long gracePeriod = GRACE_PERIOD_DEFAULT;
```
Comment (tishun) [inline src/main/java/io/lettuce/core/failover/MultiDbOptions.java]: One option to avoid wrong interpretations of what this is is to use the [Duration](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html) class:
- it carries inside of it what the duration is (milliseconds/seconds/etc)
- it could be set to null (for disabled grace period)
- it is very easy to understand the default value, e.g. Duration.ofSeconds(30)

File: src/main/java/io/lettuce/core/failover/RedisDatabaseImpl.java [hunk 2]
```diff
@@ -120,4 +126,51 @@ public State getCircuitBreakerState() {
...
+
+    /**
+     * Starts a grace period for this database. During the grace period, this database cannot be selected for failover or
+     * failback.
+     *
+     * @param durationMillis the duration of the grace period in milliseconds
+     */
+    void startGracePeriod(long durationMillis) {
+        if (durationMillis > 0) {
+            this.gracePeriodEndTime = System.currentTimeMillis() + durationMillis;
```
Comment (tishun) [inline src/main/java/io/lettuce/core/failover/RedisDatabaseImpl.java]: Similarly [Durations](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html) work very easily with the [Instant](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html) class and do the calculations for you

File: src/test/java/io/lettuce/core/failover/HealthCheckIntegrationTests.java [hunk 3]
```diff
@@ -208,7 +210,8 @@ void shouldUseDifferentStrategiesPerEndpoint() {
             DatabaseConfig config2 = DatabaseConfig.builder(uri2).weight(0.5f).healthCheckStrategySupplier(supplier2).build();
 
             // When: Create MultiDbClient with different strategies per endpoint
-            MultiDbClient testClient = MultiDbClient.create(java.util.Arrays.asList(config1, config2));
+            MultiDbClient testClient = MultiDbClient.create(java.util.Arrays.asList(config1, config2),
```
Comment (tishun) [inline src/test/java/io/lettuce/core/failover/HealthCheckIntegrationTests.java]: java.util.Arrays -> Arrays

AI is a plague

File: src/test/java/io/lettuce/test/NoFailback.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ *     // noFailbackClient has failback disabled
+ * }
+ * </pre>
+ *
+ * @author Ali Takavci
+ * @see LettuceExtension
+ */
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface NoFailback {
```
Comment (tishun) [inline src/test/java/io/lettuce/test/NoFailback.java:39]: Nice!

File: src/main/java/io/lettuce/core/failover/MultiDbOptions.java [hunk 5]
```diff
@@ -39,6 +42,16 @@ public long getFailbackCheckInterval() {
...
 
+    /**
+     * Returns the grace period duration that starts when a database is failed over from. During this grace period, no failbacks
+     * or failovers can trigger a switch back to that database until the duration ends.
+     *
+     * @return the grace period duration, or {@code null} if no grace period is configured
+     */
+    public long getGracePeriod() {
+        return gracePeriod;
+    }
```
Comment (ggivo) [inline src/main/java/io/lettuce/core/failover/MultiDbOptions.java:56]: TImeUnit is not clear

File: src/main/java/io/lettuce/core/failover/RedisDatabaseImpl.java [hunk 6]
```diff
@@ -120,4 +126,51 @@ public State getCircuitBreakerState() {
...
+     * Checks if this database is currently in a grace period.
+     *
+     * @return {@code true} if the database is in a grace period, {@code false} otherwise
+     */
+    boolean isInGracePeriod() {
+        long endTime = this.gracePeriodEndTime;
+        if (endTime == 0) {
+            return false;
+        }
+        boolean inGracePeriod = System.currentTimeMillis() < endTime;
```
Comment (ggivo) [inline src/main/java/io/lettuce/core/failover/RedisDatabaseImpl.java]: Consider using injectable java.time.Clock
Will make time based test not rely on thread sleeps

File: src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            // When: Start grace period with 100ms
+            database.startGracePeriod(100L);
+            assertThat(database.isInGracePeriod()).isTrue();
+
+            // Wait for grace period to expire
+            Thread.sleep(150L);
+
+            // Then: Database should not be in grace period
+            assertThat(database.isInGracePeriod()).isFalse();
+        }
```
Comment (ggivo) [inline src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java:127]: `Thread.sleep` in test tends to build up and slow down tests
related to the previous comment for using injectable `java.time.Clock` and modifiable Clock for the tests.

File: src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+            // Given: Circuit breaker is OPEN
+            circuitBreaker.transitionTo(CircuitBreaker.State.OPEN);
+            assertThat(circuitBreaker.getCurrentState()).isEqualTo(CircuitBreaker.State.OPEN);
+
+            // When: Start grace period with 100ms
+            database.startGracePeriod(100L);
+            assertThat(database.isInGracePeriod()).isTrue();
+
+            // Wait for grace period to expire
+            Thread.sleep(150L);
```
Comment (ggivo) [inline src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java]: same here

File: src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+            when(healthCheck1.getStatus()).thenReturn(HealthStatus.HEALTHY);
+            circuitBreaker1.transitionTo(CircuitBreaker.State.CLOSED);
+            redis1.startGracePeriod(100L);
+
+            // Then: redis1 should not be healthy initially
+            assertThat(redis1.isHealthy()).isFalse();
+
+            // When: Wait for grace period to expire
+            Thread.sleep(150L);
```
Comment (ggivo) [inline src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java]: same here

Comment (ggivo) [inline src/test/java/io/lettuce/core/failover/RedisDatabaseImplGracePeriodUnitTests.java]: one more

File: src/test/java/io/lettuce/core/failover/GracePeriodIntegrationTests.java [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+                assertThat(testConnection.getCurrentEndpoint()).isEqualTo(redis2Uri);
+            });
+
+            // When: Make redis1 healthy again
+            testStrategy.setHealthStatus(redis1Uri, HealthStatus.HEALTHY);
+
+            // Then: Should NOT failback to redis1 during grace period (3 seconds)
+            // Wait 1.5 seconds (half the grace period) and verify still on redis2
+            try {
+                Thread.sleep(1500L);
```
Comment (ggivo) [inline src/test/java/io/lettuce/core/failover/GracePeriodIntegrationTests.java]: not that easy to resolve as in unit tests.

A possible future improvement would be introducing a non time-based grace period to make this logic easier to test deterministically.
